compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchWithTransformDialect.cpp

```diff
@@ -0,0 +1,105 @@
+// Copyright 2020 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree-dialects/Dialect/LinalgExt/TransformOps/LinalgExtTransformOps.h"
+#include "iree/compiler/Dialect/Flow/IR/FlowDialect.h"
+#include "iree/compiler/Dialect/Flow/Transforms/PassDetail.h"
+#include "iree/compiler/Dialect/Flow/Transforms/Passes.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/SourceMgr.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Arithmetic/IR/Arithmetic.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/PDL/IR/PDL.h"
+#include "mlir/Dialect/PDLInterp/IR/PDLInterp.h"
+#include "mlir/Dialect/SCF/SCF.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Dialect/Transform/IR/TransformDialect.h"
+#include "mlir/Parser/Parser.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Support/FileUtilities.h"
+#include "mlir/Support/LLVM.h"
+
+static llvm::cl::opt<std::string> clTransformFileName(
```

Comment:
It'd be good to make this a pass option instead as then reproducers/etc work with it and it's scoped to the pass.
Then the -iree-flow- flag for enabling the pass could change to a string that is passed in to set the option and simplify the code paths (since it's not valid to use the flag without a file set and doesn't make sense to accept the filename param unless the flag is set).

```
--iree-flow-dispatch-use-transform-dialect --dispatch-transform-dialect-file-name=%p/transform_dialect_dispatch_spec.mlir 
```
->
```
--iree-flow-dispatch-use-transform-dialect=%p/transform_dialect_dispatch_spec.mlir 
```

Reply:
I've got some ideas on how this can come together better for such reproducers and be hermetic. Early in the pipeline, we probably want to read the file and inline it into the module, then using the module bound transformations from then on...

Reply:
(for later)

Reply:
`we probably want to read the file and inline it into the module, then using the module bound transformations from then on` -> this was the first thing I tried a while back but it really wasn't clear how to transport this module properly in the IR the the proper place an through IREE transformations and IR massaging. The file was the only sane thing I could come up with at the time. Maybe it's time to revisit but it will prob require deeper design.

Reply:
capturing some of this in my batch reply and resolving this instance.

Reply:
@benvanik it works for the e2e test but I am stuck with passing the proper pass options to `compiler/src/iree/compiler/Dialect/Flow/Transforms/test/dispatch_linalg_transform_dialect.mlir` to make an iree-opt unit test run.

Before I had:
```
// RUN: iree-opt --dispatch-transform-dialect-file-name=%p/transform_dialect_dispatch_spec.mlir \
// RUN:   --pass-pipeline="func.func(iree-flow-dispatch-linalg-on-tensors-transform-dialect-pass), cse, canonicalize, cse" 
```

Now I need:
```
// RUN: iree-opt --iree-flow-dispatch-use-transform-dialect=%p/transform_dialect_dispatch_spec.mlir \
// RUN:   --pass-pipeline="???" 
```

Could you help find the right incantation?

Reply:
I think that's the high order bit for me: of going beyond testing/experimentation, this needs more thought/discussion and I'm fine with something simple at this stage.

If river's resources stuff works out, we may have a path to just bundle it for reproducers without messing with everything else. But that is future.

Reply:
After more spelunking, passing shared IR to the transform dialect and PDL is not a viable solution. The underlying issue is that any IR manipulation in LLVM that performs a move of an op (e.g. splice) modifies **surrounding** IR (i.e. the links of the op into the surrounding block) in a non-threadsafe way.

In the longer term, extracting transform IR from a larger database and attaching it to the place where it is applied in the IR is likely the way to go.

For now, parsing is done at initialization time and cloning at `runOperation` time, as this would be a good place to perform policy filtering based on actual IR. 
Alternatively, we may also want to just clone the transform IR on each thread; this will not have filtering capabilities but OTOH such filtering will likely be its own orthogonal mechanism.

Please lmk if you have a strong preference one way of another, in any case this will disappear in some future and other mechanisms will be needed to select transform IR.